### ISSUE TITLE
Support GET /payments/:id

### ISSUE THREAD

**Issue by emschwartz** [interledger/five-bells-connector#77]
Connectors should store the details of payments they've facilitated and allow clients to query them

**Comment by vhpoet** [comment]
This is a pretty important feature. As a receiver I don't currently have a way to see who's the sender for interledger payments, all I see is the market maker in my ledger. Is there an ETA for this? @lumberj @gip

**Comment by gip** [comment]
The current plan is not to have any state/DB in the connector itself and leave that to the connector backend (which will very probably store payments and rates). I am not sure the connector should be made responsible for tracking payers/payee. Ideally the logic that initiates the payment should be doing that I'd think?

**Comment by clark800** [comment]
I think I agree with Gilles. If we think of the connector as an independent liquidity provider (market maker), they have no incentive to provide this service. Since this is an open protocol, rational profit-maximizing connectors would disable such an endpoint. Even if it would be convenient for some applications, it doesn't seem like a good idea to design the architecture around an economically fragile assumption like this.

**Comment by vhpoet** [comment]
@gip @clark800 Your point makes a lot of sense for a world where the whole infrastructure is well defined and setup, but for the short term I as a receiver don't have a way to track the sender. I'm looking at this from the product side. I'm building a UI for five-bells-ledger, and as far as I know there's currently no way to show the sender name in the interledger payment history for the UI user. My understanding is that connector already has all the information right now, and all it needs to do is store it and than remove that functionality when we have a replacement (_connector backend_). Let me know if you think there's a better place than connector to store this information in short term.

tl;dr
![image](https://cloud.githubusercontent.com/assets/686185/13299900/1e193a98-daf2-11e5-990e-c22eb6a1612c.png)

**Comment by clark800** [comment]
There is an "additional_info" field on transfers that contains "part_of_payment". How about adding the sender's ID to the "additional_info" field?

**Comment by emschwartz** [comment]
I think the connector is definitely going to need to keep its own history.

For banks they'll want all of that type of information to be stored in case anything ever goes wrong or for auditing purposes.

For Universal mode it's essential for connectors to keep state in case the connector server crashes during the escrow window for transfers they are involved in. If this happens, the connector may get a notification later that a transfer (the destination transfer) has been executed and they won't know what to do with that. At minimum the records that are currently stored [here](https://github.com/interledger/five-bells-connector/blob/master/src/services/destinationSubscriptions.js#L6) and [here](https://github.com/interledger/five-bells-connector/blob/master/src/services/sourceSubscriptions.js#L3) need to be made persistent.

**Comment by gip** [comment]
How about a very simple additional component for your demo - it would store (payer, payee, amount, status) into a db table. Two endpoints, one to add one to query. 
That would be closer to a SOA approach and could be as quick to implement without bloating the connector. I can do that quickly if you need help on that.

**Comment by emschwartz** [comment]
If it's strictly for the demo then the quickest, dirtiest way of doing it would be to have the connector store all of the payments its part of in memory. If this is necessary I'd advocate doing this on a branch that doesn't get merged in.

For anything longer term it should be a database and it should store all the proper information

**Comment by gip** [comment]
@emschwartz agreed but that could be at the connector backend level.

**Comment by emschwartz** [comment]
I think it would make more sense to do it at the connector level. What's the point of having the connector if all of this type of functionality is going to be pushed off on to some other component. The intention with the backends was just to use them to get the rates. They don't deal with who money is being sent to/from. All of that information lives with the connector so I think it would make more sense for the connector to store it

**Comment by clark800** [comment]
> For banks they'll want all of that type of information to be stored in case anything ever goes wrong or for auditing purposes.

Even if the connector stores the information, they may not want to store it forever and they may not want to provide that information to other parties for free.

> For Universal mode it's essential for connectors to keep state in case the connector server crashes during the escrow window for transfers they are involved in.

Again, the connector could still delete records after everything is finalized.

I'm not trying to say that we can't store history in the connector, just that I don't think the protocol should rely on it being available for it's basic operation.

**Comment by emschwartz** [comment]
You're right that they may not store it forever and that's fine. I'd guess that the benefit of being able to audit the history would outweigh the cost of storage, but we'll see. I think Vahe wants to see the history for demo purposes right now. I think we can address the db size issue when it arises.

The protocol does not rely on the connector persisting payments after they are completed. In the Universal mode the connector itself relies on the storage for its basic operation and ensuring that it doesn't lose money if its server goes offline temporarily. It'll also need to use a database rather than storing the subscription details in memory to allow it to scale beyond a single process.

**Comment by naoitoi** [comment]
I like @clark800 's suggestion:

> There is an "additional_info" field on transfers that contains "part_of_payment". How about adding the sender's ID to the "additional_info" field?

So @vhpoet would:
1. Look up transfer object at the receiving ledger.
2. Look up additional_info field and find the Sender's ID in (1).
3. Look up the sender by (2) at sending ledger.

Would this work for @vhpoet ?

**Comment by justmoon** [comment]
Agree with @clark800:

> There is an "additional_info" field on transfers that contains "part_of_payment". How about adding the sender's ID to the "additional_info" field?

There are serious privacy problems with the connector returning GET /payments/:id to anyone and the connector currently has no concept of authentication. (And I would be weakly against adding a lot of complexity in that direction.) 

Best solution is not to change the connector right now, but make sure the endpoints exchange whatever information they need through `memo`, `additional_info` etc. (This is the protocol level above ILP.)

Note that recipients can reject any ILP payments they don't expect (at least once we add the debit auth feature to the ledger), so they can reject any incoming transactions that don't have valid metadata.

Closing this issue. Won't fix.
